### PR TITLE
Fix deadlock between mod_muc_{room,log}.

### PR DESCRIPTION
**Pull request by kzemek**
mod_muc_room might be calling mod_muc_log:check_access_log at the
same time mod_muc_log process called mod_muc_room:get_room_users.
This commit changes the communication between processes to be
unidirectional.

### CODE REVIEW

File: apps/ejabberd/src/mod_muc_log.erl [hunk 1]
```diff
@@ -711,10 +745,11 @@ fw(F, S, FileFormat) ->
...
-        file_format()) -> 'ok'.
-put_header(_, _, _, _, _, _, _, _, _, plaintext) ->
+        file_format(), OccupantsMap :: #{binary() => [jid_nick_role()]}) -> 'ok'.
+put_header(_, _, _, _, _, _, _, _, _, plaintext, _) ->
     ok;
-put_header(F, Room, Date, CSSFile, Lang, Hour_offset, Date_prev, Date_next, Top_link, FileFormat) ->
+put_header(F, Room, Date, CSSFile, Lang, Hour_offset, Date_prev, Date_next, Top_link, FileFormat,
+           OccupantsMap) ->
     fw(F, <<"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">">>),
     fw(F, <<"<html xmlns=\"http://www.w3.org/1999/xhtml\" xml:lang=\"", Lang/binary, "\" lang=\"", Lang/binary, "\">">>),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_log.erl]: According to **Elvis**:
> Line 754 is too long:     fw(F, <<"<html xmlns=\"http://www.w3.org/1999/xhtml\" xml:lang=\"", Lang/binary, "\" lang=\"", Lang/binary, "\">">>),.

File: apps/ejabberd/src/mod_muc_log.erl [hunk 2]
```diff
@@ -711,10 +745,11 @@ fw(F, S, FileFormat) ->
 -spec put_header(file:io_device(), Room :: #room{}, Date :: binary(),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_log.erl]: According to **Elvis**:
> The spec in line 745 uses a record, please define a type for the record and use that instead.

File: apps/ejabberd/src/mod_muc_log.erl [hunk 3]
```diff
@@ -711,10 +745,11 @@ fw(F, S, FileFormat) ->
...
         Date_prev :: binary(), Date_next :: binary(), Top_link :: tuple(),
-        file_format()) -> 'ok'.
-put_header(_, _, _, _, _, _, _, _, _, plaintext) ->
+        file_format(), OccupantsMap :: #{binary() => [jid_nick_role()]}) -> 'ok'.
+put_header(_, _, _, _, _, _, _, _, _, plaintext, _) ->
     ok;
-put_header(F, Room, Date, CSSFile, Lang, Hour_offset, Date_prev, Date_next, Top_link, FileFormat) ->
+put_header(F, Room, Date, CSSFile, Lang, Hour_offset, Date_prev, Date_next, Top_link, FileFormat,
+           OccupantsMap) ->
     fw(F, <<"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">">>),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_log.erl]: According to **Elvis**:
> Line 753 is too long:     fw(F, <<"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">">>),.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 4]
```diff
@@ -3461,6 +3461,7 @@ get_config(Lang, StateData, From) ->
         {N, integer_to_binary(N)};
         _ -> {0, <<"none">>}
     end,
+
     Res =
     [#xmlel{name = <<"title">>,
             children = [#xmlcdata{content = <<(translate:translate(Lang, <<"Configuration of room ">>))/binary,
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Line 3467 is too long:             children = [#xmlcdata{content = <<(translate:translate(Lang, <<"Configuration of room ">>))/binary,.

File: apps/ejabberd/src/mod_muc_log.erl [hunk 5]
```diff
@@ -711,10 +746,11 @@ fw(F, S, FileFormat) ->
...
     fw(F, <<"<html xmlns=\"http://www.w3.org/1999/xhtml\" xml:lang=\"", Lang/binary, "\" lang=\"", Lang/binary, "\">">>),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_log.erl]: According to **Elvis**:
> Line 755 is too long:     fw(F, <<"<html xmlns=\"http://www.w3.org/1999/xhtml\" xml:lang=\"", Lang/binary, "\" lang=\"", Lang/binary, "\">">>),.

File: apps/ejabberd/src/mod_muc_log.erl [hunk 6]
```diff
@@ -711,10 +746,11 @@ fw(F, S, FileFormat) ->
...
         Date_prev :: binary(), Date_next :: binary(), Top_link :: tuple(),
-        file_format()) -> 'ok'.
-put_header(_, _, _, _, _, _, _, _, _, plaintext) ->
+        file_format(), OccupantsMap :: #{binary() => [jid_nick_role()]}) -> 'ok'.
+put_header(_, _, _, _, _, _, _, _, _, plaintext, _) ->
     ok;
-put_header(F, Room, Date, CSSFile, Lang, Hour_offset, Date_prev, Date_next, Top_link, FileFormat) ->
+put_header(F, Room, Date, CSSFile, Lang, Hour_offset, Date_prev, Date_next, Top_link, FileFormat,
+           OccupantsMap) ->
     fw(F, <<"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">">>),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_log.erl]: According to **Elvis**:
> Line 754 is too long:     fw(F, <<"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">">>),.

File: apps/ejabberd/src/mod_muc_log.erl [hunk 7]
```diff
@@ -711,10 +746,11 @@ fw(F, S, FileFormat) ->
 -spec put_header(file:io_device(), Room :: #room{}, Date :: binary(),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_log.erl]: According to **Elvis**:
> The spec in line 746 uses a record, please define a type for the record and use that instead.

**Comment by michalwski** [comment]
Thanks!

**Pull request merged**